### ISSUE TITLE
Context API: Consider method to update context value depending on emitted elements

### ISSUE THREAD

**Issue by simonbasle** [reactor/reactor-core#699]
`contextMap` allows to enrich the context at subscription time, but maybe we need an API to enrich the context and update it depending on received values, ie values emitted by upstream.

Something like `contextPut(Object key, Function<T, ?> valueExtractor)`.

**Comment by simonbasle** [comment]
do you suspect you'll need that @rstoyanchev @rwinch ?

**Comment by rstoyanchev** [comment]
It seems useful but I don't have a case. I can say a dynamic, Function-backed value feels a bit at odds with the mental image of a Map which is what this looks like so far.

**Comment by smaldini** [comment]
The context is immutable, doing put returns a new instance , hence why current contextMap must return a Context (in case you wanted to mutate it). 
What I'm thinking more and more is that we should not authorize any update at all for the context but the initial contextMap that enriches it at subscription time. If someone wants to connect an upstream to a downstream it can use its own state inside the context holder but we would just keep doing the bare minimum e.g. : 

```
class ValueHolder{ int v };

flux.contextMap(ctx -> {
   ctx.get(ValueHolder.class).value = 2;
   return ctx;
})
 .doOnNext(xxx)
 .map(xx)
//...
  .contextMap(ctx -> ctx.put(ValueHolder.class, new ValueHolder()))
  .subscribe()

```

**Comment by smaldini** [comment]
Will not fix it since #704 is actually generic for now and the internal might not support well this kind of reinjecting. I would advise users to setup their context at first to support this with e.g. a value holder or even a contextualized mono/flux (e.g. acknowledgements, req/reply...).
